test/unit/cfn_yaml/core_test.clj

```diff
@@ -106,3 +106,33 @@ Resources:
 (deftest cidr
   (is (= (!Cidr "192.168.0.0/24" 6 5)
          (sut/parse "!Cidr [ \"192.168.0.0/24\", 6, 5 ]"))))
+
+(deftest find-in-map
+  (is (= (!FindInMap "SubnetConfig" (!Ref "AWS::Region") "CIDR")
+         (sut/parse* "!FindInMap ['SubnetConfig', !Ref 'AWS::Region', 'CIDR']")))
+
+  (is (= (!FindInMap "SubnetConfig" "VPC" "CIDR")
+         (sut/parse "!FindInMap ['SubnetConfig', 'VPC', 'CIDR']"))))
+
+(deftest find-in-map-gen
+  (is (= "!FindInMap [\"SubnetConfig\", \"VPC\", \"CIDR\"]\n"
+         (sut/generate-string (!FindInMap "SubnetConfig" "VPC" "CIDR")))))
+
+
+(deftest get-att
+  (is (= (!GetAtt "myELB" "DNSName")
+         (sut/parse* "!GetAtt myELB.DNSName"))))
+
+(deftest get-att-gen
+  (is (= "!GetAtt 'myELB.DNSName'\n"
+         (sut/generate-string (!GetAtt "myELB" "DNSName")))))
+
+
+(deftest join-attr
+  (is (= (!Join ":" [(!Ref "AWS::StackName") "ECSTaskExecutionRole"])
+         (sut/parse* "!Join [ ':', [ !Ref 'AWS::StackName', 'ECSTaskExecutionRole']]"))))
+
+;; TODO don't know how to generate vectors
+#_(deftest join-attr-gen
```

Comment:
It's been a while I hacked this, didn't figure out myself yet 😅 Anyways, [cfn-yaml.tags/representers](https://github.com/portkey-cloud/cfn-yaml/blob/dd50a8bd3a5f6d4eed26feed276c6f94f4666c5f/src/cfn_yaml/tags.clj#L98-L131) would be the place to do it. There are some examples in existing representers. The dilemma here is to call the embedded representers correctly, based on the type of data in the !Join (See !Base64 for example).

Reply:
So seems that `!Base64` is also broken for intrinsics:

```clojure
unit.cfn-yaml.core-test> (sut/generate-string* {:UserData (!Base64 (!Ref "moi"))})
"UserData: !Base64\n  ? !!clojure.lang.Keyword\n    sym: {}\n  : moi\n"
```
So need more diving in on how nested representation works, might be good to read snakeyaml documentation a bit :)

Reply:
So I think this requires recursion, passing the `Representer` object for use in our custom representers seems to work:

```
diff --git a/src/cfn_yaml/core.clj b/src/cfn_yaml/core.clj
index 2d4d32a..b617f0c 100644
--- a/src/cfn_yaml/core.clj
+++ b/src/cfn_yaml/core.clj
@@ -62,7 +62,8 @@
     (.putAll (.get representers-field representer)
              (tags/representers #(.invoke represent-data
                                           representer
-                                          (into-array Object [%]))))
+                                          (into-array Object [%]))
+                                representer))
     yaml))
 
 (defn find-references [template]
diff --git a/src/cfn_yaml/tags.clj b/src/cfn_yaml/tags.clj
index 5945075..1d4015f 100644
--- a/src/cfn_yaml/tags.clj
+++ b/src/cfn_yaml/tags.clj
@@ -95,7 +95,7 @@
                nil
                style))
 
-(defn representers [represent-data]
+(defn representers [represent-data representer]
   (let [represent-map (fn [m & {:keys [tag] :or {tag Tag/MAP}}]
                         (MappingNode. tag
                                       (for [[k v] m]
@@ -115,6 +115,13 @@
                                   (scalar-node Tag/INT (str (:count %)))
                                   (scalar-node Tag/INT (str (:cidrBits %)))]
                                  DumperOptions$FlowStyle/FLOW)]
+          [!Join #(SequenceNode. (Tag. "!Join")
+                                 (map (fn [x]
+                                        (if (string? x)
+                                          (scalar-node Tag/STR x :style DumperOptions$ScalarStyle/DOUBLE_QUOTED)
+                                          (.represent representer x)))
+                                      (:list-of-values %))
+                                 DumperOptions$FlowStyle/FLOW)]
           [!FindInMap #(SequenceNode. (Tag. "!FindInMap")
                                       [(scalar-node Tag/STR (:mapName %) :style DumperOptions$ScalarStyle/DOUBLE_QUOTED)
                                        (scalar-node Tag/STR (:topLevelKey %) :style DumperOptions$ScalarStyle/DOUBLE_QUOTED)
@@ -122,9 +129,9 @@
                                       DumperOptions$FlowStyle/FLOW)]
           [!GetAtt #(scalar-node "!GetAtt" (str (:logicalNameOfResource %) "." (:attributeName %)))]
           [!Base64 (fn [{:keys [valueToEncode]}]
-                     (cond
-                       (string? valueToEncode) (scalar-node "!Base64" valueToEncode)
-                       (map? valueToEncode) (represent-map valueToEncode :tag (Tag. "!Base64"))))]]
+                     (if (string? valueToEncode)
+                       (scalar-node "!Base64" valueToEncode)
+                       (.represent representer valueToEncode)))]]
          (into {} (map (fn [[klass f]]
                          [klass (reify org.yaml.snakeyaml.representer.Represent
                                   (representData [this data]
```

This yields:

```clojure
unit.cfn-yaml.core-test> (sut/generate-string* (!Join ":" [(!Ref "AWS::StackName") "ECSTaskExecutionRole"]))
"!Join [!Ref 'AWS::StackName', \"ECSTaskExecutionRole\"]\n"
```

I'm a little hazy on the 

```clojure
(if (string? valueToEncode)
                       (scalar-node "!Base64" valueToEncode) ...
```

, but probably can live with that for now :)